### PR TITLE
ecs: api/doc changes for merging spellcasting to class

### PR DESCRIPTION
**Pull request by ecshreve**
## What does this do?
- updates docs to reflect changes in https://github.com/bagelbits/5e-database/pull/316
- move spellcasting data under the associated class
- remove all top level references to spellcasting

## How was it tested?
- localhost
- ran the db/api locally and issued `curl` requests to various `/class` endpoints and validated the responses

example

`curl http://localhost:3000/api/classes/warlock | python -m json.tool`

```bash
{
    "index": "warlock",
    "name": "Warlock",
    "hit_die": 8,
    "proficiency_choices": [...],
    "proficiencies": [...],
    "starting_equipment": "/api/starting-equipment/warlock",
    "class_levels": "/api/classes/warlock/levels",
    "subclasses": [...],
    "spellcasting": {
        "level": 1,
        "spellcasting_ability": {
            "index": "cha",
            "name": "CHA",
            "url": "/api/ability-scores/cha"
        },
        "info": [
            {
                "name": "Cantrips",
                "desc": [
                    "You know two cantrips of your choice from the warlock spell list. You learn additional warlock cantrips of your choice at higher levels, as shown in the Cantrips Known column of the Warlock table."
                ]
            },
            {
                "name": "Spell Slots",
                "desc": [
                    "The Warlock table shows how many spell slots you have. The table also shows what the level of those slots is; all of your spell slots are the same level. To cast one of your warlock spells of 1st level or higher, you must expend a spell slot. You regain all expended spell slots when you finish a short or long rest.",
                    "For example, when you are 5th level, you have two 3rd-level spell slots. To cast the 1st-level spell thunderwave, you must spend one of those slots, and you cast it as a 3rd-level spell."
                ]
            },
            {
                "name": "Spells Known of 1st Level and Higher",
                "desc": [
                    "At 1st level, you know two 1st-level spells of your choice from the warlock spell list.",
                    "The Spells Known column of the Warlock table shows when you learn more warlock spells of your choice of 1st level and higher. ",
                    "A spell you choose must be of a level no higher than what's shown in the table's Slot Level column for your level. When you reach 6th level, for example, you learn a new warlock spell, which can be 1st, 2nd, or 3rd level.",
                    "Additionally, when you gain a level in this class, you can choose one of the warlock spells you know and replace it with another spell from the warlock spell list, which also must be of a level for which you have spell slots."
                ]
            },
            {
                "name": "Spellcasting Ability",
                "desc": [
                    "Charisma is your spellcasting ability for your warlock spells, so you use your Charisma whenever a spell refers to your spellcasting ability. In addition, you use your Charisma modifier when setting the saving throw DC for a warlock spell you cast and when making an attack roll with one."
                ]
            },
            {
                "name": "Spellcasting Focus",
                "desc": [
                    "You can use an arcane focus as a spellcasting focus for your warlock spells."
                ]
            }
        ]
    },
    "spells": "/api/classes/warlock/spells",
    "url": "/api/classes/warlock"
}
```

- just the spellcasting portion
`curl http://localhost:3000/api/classes/warlock/spellcasting | python -m json.tool`

```bash
{
    "level": 1,
    "spellcasting_ability": {
        "index": "cha",
        "name": "CHA",
        "url": "/api/ability-scores/cha"
    },
    "info": [
        {
            "name": "Cantrips",
            "desc": [
                "You know two cantrips of your choice from the warlock spell list. You learn additional warlock cantrips of your choice at higher levels, as shown in the Cantrips Known column of the Warlock table."
            ]
        },
        {
            "name": "Spell Slots",
            "desc": [
                "The Warlock table shows how many spell slots you have. The table also shows what the level of those slots is; all of your spell slots are the same level. To cast one of your warlock spells of 1st level or higher, you must expend a spell slot. You regain all expended spell slots when you finish a short or long rest.",
                "For example, when you are 5th level, you have two 3rd-level spell slots. To cast the 1st-level spell thunderwave, you must spend one of those slots, and you cast it as a 3rd-level spell."
            ]
        },
        {
            "name": "Spells Known of 1st Level and Higher",
            "desc": [
                "At 1st level, you know two 1st-level spells of your choice from the warlock spell list.",
                "The Spells Known column of the Warlock table shows when you learn more warlock spells of your choice of 1st level and higher. ",
                "A spell you choose must be of a level no higher than what's shown in the table's Slot Level column for your level. When you reach 6th level, for example, you learn a new warlock spell, which can be 1st, 2nd, or 3rd level.",
                "Additionally, when you gain a level in this class, you can choose one of the warlock spells you know and replace it with another spell from the warlock spell list, which also must be of a level for which you have spell slots."
            ]
        },
        {
            "name": "Spellcasting Ability",
            "desc": [
                "Charisma is your spellcasting ability for your warlock spells, so you use your Charisma whenever a spell refers to your spellcasting ability. In addition, you use your Charisma modifier when setting the saving throw DC for a warlock spell you cast and when making an attack roll with one."
            ]
        },
        {
            "name": "Spellcasting Focus",
            "desc": [
                "You can use an arcane focus as a spellcasting focus for your warlock spells."
            ]
        }
    ]
}
```

- verified docs locally
Screen Shot 2020-11-02 at 05 27 35 AM
<hr>
Screen Shot 2020-11-02 at 05 27 45 AM

<hr>

Screen Shot 2020-11-05 at 3 49 42 AM

## Is there a Github issue this is resolving?
- this addresses one part of https://github.com/bagelbits/5e-database/issues/251

## Here's a fun image for your troubles
random photo - update me

### CODE REVIEW

File: src/routes/api/classes.js [hunk 1]
```diff
@@ -11,7 +11,6 @@ router.get('/:index', function(req, res, next) {
 
 router.get('/:index/subclasses', ClassController.showSubclassesForClass);
 router.get('/:index/starting-equipment', ClassController.showStartingEquipmentForClass);
-router.get('/:index/spellcasting', ClassController.showSpellcastingForClass);
```
Comment (bagelbits) [inline src/routes/api/classes.js:14]: I thought we were still keeping this endpoint but only having it grab from the class object?

File: src/views/partials/doc-content.ejs [hunk 2]
```diff
@@ -126,8 +126,7 @@
 
   <h1 id="classes-section">Classes</h1>
   <% include ../partials/doc-resource-classes %> <% include ../partials/doc-resource-subclasses %>
-  <% include ../partials/doc-resource-features %> <% include ../partials/doc-resource-spellcasting
-  %> <% include ../partials/doc-resource-starting-equipment %>
+  <% include ../partials/doc-resource-starting-equipment %>
```
Comment (bagelbits) [inline src/views/partials/doc-content.ejs:129]: I think you deleted features as well?

Comment (ecshreve) [reply src/views/partials/doc-content.ejs:129]: Yep I totally did, will fix

File: src/routes/api/classes.js (hunk 1 above)
Comment (ecshreve) [reply src/routes/api/classes.js:14]: Yep, yep we were, I...forgot. Let me re-work this a bit, I think it's doable.

Comment (bagelbits) [reply src/routes/api/classes.js:14]: It's probably going to fail integration tests until we get the other one merged in.

Comment (ecshreve) [reply src/routes/api/classes.js:14]: This one is going to take a little more work. Separately, I think I'll need to make some tweaks to the integration tests themselves, but yes they will probably fail integration tests in CI until the other one is merged.

**Comment by ecshreve** [comment]
@fergcb - [not urgent] - when you have a chance, I _think_ this set of changes is more or less in line with what we were talking about on discord earlier, curious to hear your thoughts on how it looks

File: src/routes/api/classes.js (hunk 1 above)
Comment (ecshreve) [reply src/routes/api/classes.js:14]: How strongly do we feel about keeping these spellcasting details accessible via their own endpoint? Asking for 2 reasons:
- it seems fairly straightforward / trivial on the user side to hit the `/classes/{index}` endpoint and pull the `spellcasting` field out from there
- right now there isn't anything accessible via it's own endpoint like this that isn't defined in it's own collection in mongodb, and the db change associated with this PR effectively removes all the documents from the `spellcasting` collection and just drops the raw data into the appropriate document in the `class` collection

I'm not too familiar with mongo/express/mongoose, so I've been doing some reading to try and work out a way to do this, but curious if maybe I'm missing something obvious that y'all could point me toward.

Comment (bagelbits) [reply src/routes/api/classes.js:14]: I don't feel super strong about it? Just slight backwards compatibility.

Comment (fergcb) [reply src/routes/api/classes.js:14]: Same as @bagelbits. I personally am not bothered whether it stays or not, but having that bit of backwards compatibility would be nice, and there is the slight benefit of being able to get a smaller response if all you need is the spellcasting data.

I wouldn't consider "there isn't anything else like this" as a reason not to do it, as long as it has a net positive effect.

As for the obvious way of doing it, I would recommend modifying `classController.js`'s `showSpellcastingForClass` function to something like

```js
Class.findOne({ index: req.params.index })
  .then(class => {
    res.status(200).json(class.spellcasting)
  })
  .catch(err => {
    next(err)
  })
```

Comment (bagelbits) [reply src/routes/api/classes.js:14]: Also, you can probably reuse the original spellcasting integration tests and unit tests but for that endpoint. So not too much additional work.

File: src/views/partials/doc-content.ejs (hunk 2 above)
Comment (ecshreve) [reply src/views/partials/doc-content.ejs:130]: this is fixed

File: src/routes/api/classes.js (hunk 1 above)
Comment (ecshreve) [reply src/routes/api/classes.js]: yeah the tests ended up staying pretty much the same, i added them back in a subsequent commit on this PR for readability's sake, happy to squash the changes back into the commit original commit though

**Comment by ecshreve** [comment]
Okay, I think this PR is ready for another look whenever y'all have a chance.

**Review by bagelbits — APPROVED** [review]
I think that looks great!

**Comment by ecshreve** [comment]
> I think that looks great!

If @fergcb thinks this looks okay too I think it would be pretty easy to do something just like this PR for starting-equipment

**Comment by kylrs** [comment]
Absolutely. Looks good to me!

**Comment by fergcb** [comment]
@ecshreve LGTM!

**Pull request merged**